tasmota/xsns_62_esp32_mi.ino

```diff
@@ -1139,83 +1160,110 @@ void MI32ConnectionTask(void *pvParameters){
 
-      pSvc = MI32Client->getService(MI32.conCtx->serviceUUID);
-      if(pSvc) {
-          pChr = pSvc->getCharacteristic(MI32.conCtx->charUUID);
-      }
-      else{
-        MI32.conCtx->error = MI32_CONN_NO_SERVICE;
-      }
-      if (pChr){
-        switch(MI32.conCtx->operation){
-          case 1:
-            if(pChr->canRead()) {
-            std::string _val = pChr->readValue();
-            MI32.conCtx->buffer[0] = (uint8_t)_val.size();
-            const char *_c_val = _val.c_str();
-            memcpy( MI32.conCtx->buffer + 1,_c_val,MI32.conCtx->buffer[0]);
-            }
-            else{
-            MI32.conCtx->error = MI32_CONN_CAN_NOT_READ;
-            }
-            break;
-          case 2:
-            if(pChr->canWrite()) {
-              uint8_t len = MI32.conCtx->buffer[0];
-              if(pChr->writeValue(MI32.conCtx->buffer + 1,len,true)) { // true is important !
-                // AddLog(LOG_LEVEL_DEBUG,PSTR("M32: write op done"));
+      // AddLog(LOG_LEVEL_INFO,PSTR("M32: start connection loop"));
+      bool keepConnectionAlive = true;
+      MI32.mode.triggerNextConnJob = 1;
+      while(keepConnectionAlive){
+        while(MI32.mode.triggerNextConnJob == 0){
+          vTaskDelay(50/ portTICK_PERIOD_MS);
+          if(MI32.mode.connected == 0){
+              MI32StartTask(MI32_TASK_SCAN);
+              vTaskDelete( NULL );
+          }
+          // AddLog(LOG_LEVEL_INFO,PSTR("M32: wait ..."));
+        }
+        // AddLog(LOG_LEVEL_INFO,PSTR("M32: really start connection job now ..."));
+        MI32.mode.triggerNextConnJob = 0;
+        MI32.mode.readyForNextConnJob = 0;
+        pSvc = MI32Client->getService(MI32.conCtx->serviceUUID);
+        if(pSvc) {
+            pChr = pSvc->getCharacteristic(MI32.conCtx->charUUID);
+        }
+        else{
+          MI32.conCtx->error = MI32_CONN_NO_SERVICE;
+        }
+        if (pChr){
+          switch(MI32.conCtx->operation){
+            case 1:
+              if(pChr->canRead()) {
+              NimBLEAttValue _val = pChr->readValue();
+              MI32.conCtx->buffer[0] = _val.size();
+                 memcpy( MI32.conCtx->buffer + 1,_val.data(),MI32.conCtx->buffer[0]);
               }
               else{
-              MI32.conCtx->error = MI32_CONN_DID_NOT_WRITE; 
+              MI32.conCtx->error = MI32_CONN_CAN_NOT_READ;
               }
-            }
-            else{
-              MI32.conCtx->error = MI32_CONN_CAN_NOT_WRITE;
-            }
-            MI32.mode.readingDone = 1;
-            break;
-          case 3:
-            if(pChr->canNotify()) {
-              if(pChr->subscribe(true,MI32notifyCB,false)) AddLog(LOG_LEVEL_DEBUG,PSTR("M32: subscribe"));
-            }
-            else{
-              MI32.conCtx->error = MI32_CONN_CAN_NOT_NOTIFY;
-            }
-            break;
-        default:
-            break;
+              break;
+            case 2:
+              if(pChr->canWrite()) {
```

Comment:
`              if(pChr->canWrite() || (pChr->canWriteNoResponse() && !MI32.conCtx->response)) {`

This change fixed my write issue!

Reply:
I would have thought canWrite covered with response and without response and canWriteNoResponse would help you differentiate if needed but it seems they are disjoint. canWrite == canWriteWithResponse only.